clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/chunk/ChunkAssembler.java

```diff
@@ -24,129 +20,156 @@
  * message.
  */
 public class ChunkAssembler {
+  private static final OptimizedBinaryDecoderFactory OPTIMIZED_BINARY_DECODER_FACTORY =
+      OptimizedBinaryDecoderFactory.defaultFactory();
   private static final RecordDeserializer<KafkaInputMapperValue> KAFKA_INPUT_MAPPER_VALUE_AVRO_SPECIFIC_DESERIALIZER =
       FastSerializerDeserializerFactory
           .getFastAvroSpecificDeserializer(KafkaInputMapperValue.SCHEMA$, KafkaInputMapperValue.class);
-  private final AvroSpecificDeserializer<ChunkedKeySuffix> chunkedKeySuffixDeserializer;
   private final ChunkedValueManifestSerializer manifestSerializer;
 
   public ChunkAssembler() {
-    SpecificDatumReader<ChunkedKeySuffix> specificDatumReader = new SpecificDatumReader<>(ChunkedKeySuffix.class);
-    this.chunkedKeySuffixDeserializer = new AvroSpecificDeserializer<>(specificDatumReader);
     this.manifestSerializer = new ChunkedValueManifestSerializer(true);
   }
 
+  /**
+   * The `valueIterator` of this function is supposed to be in descending order by offset.
+   *
+   * Here is the high-level algo:
+   * 1. If the latest event is a `DELETE`, return;
+   * 2. If the latest event is a regular 'PUT`, return;
+   * 3. If the latest event is a manifest, capture all the chunk info from the latest one and ignore the older ones.
+   * 4. For chunks:
```

Comment:
Hmm, I am wondering, since all the chunks will be grouped into the same key here,
then shouldn't it always be the case for chunked value?
low_offset <- chunk1, ..., chunkN, manifest -> high_offset
for the 4(a) is it possible?

Reply:
Or is it just safeguard?